R/splitRaster.R

```diff
@@ -80,69 +92,139 @@ setGeneric("splitRaster", function(x, nx, ny, buffer) {
 setMethod(
   "splitRaster",
   signature = signature(x = "RasterLayer", nx = "integer",
-                        ny = "integer", buffer = "numeric"),
-  definition = function(x, nx, ny, buffer) {
+                        ny = "integer", buffer = "numeric",
+                        savePath = "character"),
+  definition = function(x, nx, ny, buffer, savePath) {
     if (length(buffer) > 2) {
       warning("buffer contains more than 2 elements - only the first two will be used.")
       buffer <- buffer[1:2]
     } else if (length(buffer) == 1) {
       buffer <- c(buffer, buffer)
     }
-
     if (buffer[1] < 1) {
       buffer[1] <- ceiling((buffer[1]*(xmax(x) - xmin(x))/nx)/xres(x))
     }
     if (buffer[2] < 1) {
       buffer[2] <- ceiling((buffer[2]*(ymax(x) - ymin(x))/ny)/yres(x))
     }
     ext <- extent(x)
-    tiles <- vector("list", length = nx*ny)
+    extents <- vector("list", length = nx*ny)
+    if(!dir.exists(savePath)) {dir.create(savePath)}
     n <- 1L
     for (i in seq_len(nx) - 1L) {
       for (j in seq_len(ny) - 1L) {
         x0 <- ext@xmin + i*((ext@xmax - ext@xmin) / nx) - buffer[1]*xres(x)
         x1 <- ext@xmin + (i + 1L)*((ext@xmax - ext@xmin) / nx) + buffer[1]*xres(x)
         y0 <- ext@ymin + j*((ext@ymax - ext@ymin) / ny) - buffer[2]*yres(x)
         y1 <- ext@ymin + (j + 1L)*((ext@ymax - ext@ymin) / ny) + buffer[2]*yres(x)
-        tiles[[n]] <- crop(x, extent(x0, x1, y0, y1))
+        extents[[n]] <- extent(x0, x1, y0, y1)
         n <- n + 1L
       }
     }
+    cl <- tryCatch(getCluster(), error = function(x) NULL)
+    on.exit(if (!is.null(cl)) returnCluster())
+    if (!is.null(cl)) {
+      parFun <- "clusterApply"
+      cropNew <- function(ras, x0 = x, ...) {
+        crop(x = ras,y = x0)
+      }
+      args <- list(cl = cl, x = extents,
+                   fun = cropNew, ras = x)
+
+      # args <- list(seq = extents, fun = crop, r)
+      # args1 <- list(cl = cl, y = extents, fun = crop, x = r)
+      # args <- append(list(cl = cl), args)
+    } else {
+      parFun <- "lapply"
+      args <- list(extents, FUN = crop, x = r)
+    }
+    tiles <- do.call(get(parFun), args)
+    for(i in 1:length(tiles)){
```

Comment:
Alex,
Thanks.


On Tue, Jun 21, 2016 at 1:31 PM, Alex Chubaty <notifications@github.com>
wrote:

> In R/splitRaster.R
> <https://github.com/PredictiveEcology/SpaDES/pull/286#discussion_r67945805>
> :
>
> > +      parFun <- "clusterApply"
> > +      cropNew <- function(ras, x0 = x, ...) {
> > +        crop(x = ras,y = x0)
> > +      }
> > +      args <- list(cl = cl, x = extents,
> > +                   fun = cropNew, ras = x)
> > +
> > +      # args <- list(seq = extents, fun = crop, r)
> > +      # args1 <- list(cl = cl, y = extents, fun = crop, x = r)
> > +      # args <- append(list(cl = cl), args)
> > +    } else {
> > +      parFun <- "lapply"
> > +      args <- list(extents, FUN = crop, x = r)
> > +    }
> > +    tiles <- do.call(get(parFun), args)
> > +    for(i in 1:length(tiles)){
>
> I have now figured out the problem. it was an issue with the mergeRaster
> examples -- they were all using the same files to write the tiles, so they
> were overwriting each other.
>
> I will push the cleaned up PR and merge as soon as I confirm that all
> tests pass.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PredictiveEcology/SpaDES/pull/286/files/a733cb411ca2c41632018cd6f2372a7bd13146e0#r67945805>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AJ-1fEA77JuNK-P9642Yi2lq1zu-NxH_ks5qOEomgaJpZM4I3FuJ>
> .
>



-- 
Yong Luo
Postdoctoral Researcher

Canadian Forest Service
Natural Resources Canada
Pacific Forestry Centre
506 Burnside Road W
Victoria, BC  V8Z 1M5
Canada
phone: +1-250-298-2566
&
Faculty of Forestry
University of British Columbia
2424 Main Mall
Vancouver, BC V6T 1Z4
Canada